Module splitting

tlively opened the pull request:
Adds the capability to programatically split a module into a primary and
secondary module such that the primary module can be compiled and run before the
secondary module has been instantiated. All calls to secondary functions (i.e.
functions that have been split out into the secondary module) in the primary
module are rewritten to be indirect calls through the table. Initially, the
table slots of all secondary functions contain references to imported
placeholder functions. When the secondary module is instantiated, it will
automatically patch the table to insert references to the original functions.

The process of module splitting involves these steps:

 1. Create the new secondary module.

 2. Export globals, events, tables, and memories from the primary module and
    import them in the secondary module.

 3. Move the deferred functions from the primary to the secondary module.

 4. For any secondary function exported from the primary module, export in
    its place a trampoline function that makes an indirect call to its
    placeholder function (and eventually to the original secondary function),
    allocating a new table slot for the placeholder if necessary.

 5. Rewrite direct calls from primary functions to secondary functions to be
    indirect calls to their placeholder functions (and eventually to their
    original secondary functions), allocating new table slots for the
    placeholders if necessary.

 6. For each primary function directly called from a secondary function, export
    the primary function if it is not already exported and import it into the
    secondary module.

 7. Replace all references to secondary functions in the primary module's table
    segments with references to imported placeholder functions.

 8. Create new active table segments in the secondary module that will replace
    all the placeholder function references in the table with references to
    their corresponding secondary functions upon instantiation.

Functions can be used or referenced three ways in a WebAssembly module: they can
be exported, called, or placed in a table. The above procedure introduces a
layer of indirection to each of those mechanisms that removes all references to
secondary functions from the primary module but restores the original program's
semantics once the secondary module is instantiated. As more mechanisms that
reference functions are added in the future, such as ref.func instructions, they
will have to be modified to use a similar layer of indirection.

The code as currently written makes a few assumptions about the module that is
being split:

 1. It assumes that mutable-globals is allowed. This could be worked around by
    introducing wrapper functions for globals and rewriting secondary code that
    accesses them, but now that mutable-globals is shipped on all browsers,
    hopefully that extra complexity won't be necessary.

 2. It assumes that all table segment offsets are constants. This simplifies the
    generation of segments to actively patch in the secondary functions without
    overwriting any other table slots. This assumption could be relaxed by 1)
    having secondary segments re-write primary function slots as well, 2)
    allowing addition in segment offsets, or 3) synthesizing a start function to
    modify the table instead of using segments.

 3. It assumes that each function appears in the table at most once. This isn't
    necessarily true in general or even for LLVM output after function
    deduplication. Relaxing this assumption would just require slightly more
    complex code, so it is a good candidate for a follow up PR.

Future Binaryen work for this feature includes providing a command line tool
exposing this functionality as well as C API, JS API, and fuzzer support. We
will also want to provide a simple instrumentation pass for finding dead or
late-executing functions that would be good candidates for splitting out. It
would also be good to integrate that instrumentation with future function
outlining work so that dead or exceptional basic blocks could be split out into
a separate module.

sbc100 commented on src/wasm-builder.h:
```diff
@@ -221,8 +221,9 @@ class Builder {
   }
+  template<typename T>
   CallIndirect* makeCallIndirect(Expression* target,
-                                 const std::vector<Expression*>& args,
```
What else can this be?

tlively replied:
In this case it is an arena-allocated `ExpressionList`, used when I copy a `Call`'s arguments to a new `CallIndirect` expression. It would be more complicated and less efficient to copy the arguments out into a vector, just to have them copied back into an `ExpressionList`.

sbc100 commented on src/wasm.h:
```diff
@@ -1734,10 +1734,10 @@ class Module {
+  Export* addExport(std::unique_ptr<Export>&& curr);
+  Function* addFunction(std::unique_ptr<Function>&& curr);
+  Global* addGlobal(std::unique_ptr<Global>&& curr);
+  Event* addEvent(std::unique_ptr<Event>&& curr);
```
Why this?

tlively replied:
I wanted to use the overloads of these functions that take `std::unique_ptr` rather than raw pointers, since allocating an object with `std::make_unique` up front is more efficient than wrapping a raw pointer into a `std::unique_ptr`. However, `std::unique_ptr` can't be copied, so it wasn't actually possible to call these overloads the way I needed to without changing them to take rvalue references.

sbc100 commented on test/example/module-splitting.cpp:
```diff
@@ -0,0 +1,208 @@
+
+int main() {
+  // Trivial module
+  do_test({}, "(module)");
```
Do we not any anything better than hand-rolling for C++ unit tests in binaryen?  i.e. no gtest?

tlively replied:
Unfortunately not. I wish we had gtest (and LLVM's lit and FileCheck), but overhauling all the testing infrastructure would be a big project. I wouldn't want to do it only half way and leave us with yet another way to run tests. I already accidentally did that when I introduced the Python unit tests.

sbc100 reviewed (commented):
Very impressive!   

I've not finished reading the actual implementation yet but I did have one question that I'm not sure is covered by your description:

What happens when the secondary module wants to access on of the imports of the original module.  e.g. call the `env.sleep` import?   Do those imports get duplicated in the secondary module, or re-exported from the primary to the secondary under a different/internal name?

tlively commented:
> What happens when the secondary module wants to access on of the imports of the original module. e.g. call the `env.sleep` import? Do those imports get duplicated in the secondary module, or re-exported from the primary to the secondary under a different/internal name?

The import should be re-exported to be imported into the secondary module. My general principle is that the primary module should export anything that needs to be imported into the secondary module. That should simplify the linking of the two, since it lets you pass only the exports of the primary module as the imports to the secondary module with not extra application-specific logic.

I don't think I have a test for that case yet, though. I will add one.

aheejin commented:
Can you link your presentation slide you used in the SOIL seminar here? I forgot some of the rationales of the design you mentioned there, so..

tlively commented:
Sure, here are the [slides](https://docs.google.com/presentation/d/1lBtUWP0M9QDa73O4WsTBRtXlofqguz7XV4xW54lsI2M/edit?usp=sharing) and here is the [recording](https://cdnapisec.kaltura.com/html5/html5lib/v2.84.1/mwEmbedFrame.php/p/520801/uiconf_id/31230141/entry_id/1_rob8glb6?wid=_520801&iframeembed=true&playerId=kaltura_player&entry_id=1_rob8glb6&flashvars[streamerType]=auto). I also need to write a design doc for this and share it with the wider team, which I probably should have done first except it was more fun to just write some code 😬

sbc100 commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+    if (exportIt != exportedPrimaryFuncs.end()) {
+      exportName = exportIt->second;
+    } else {
+      exportName = Names::getValidExportName(primary, primaryFunc);
```
I wonder if we should use a unique / configurable prefix for the exports do distinguish between the actual function exports and the exports that are used for module splitting?   I'd even suggest that maybe we should ignore any existing exports and export always with our specially-named exports.

We could even pick a reasonable default mangling scheme that doesn't collide with C symbol namespace?  How about "%" as the prefix?  Or maybe that is too cute.

tlively replied:
Is the worry that users will accidentally depend on the new exports? I'm not sure how serious of a problem that would be. Even if we did use a special prefix, there's nothing stopping the same prefix being used in the input module, so I'm not sure this would buy us much. Maybe it would make the tests more readable?

sbc100 replied:
I was thinking for a tool like emscripten we might want to use a prefix (and it would be easy for us to choose a collision-free one) so that we could treat the exports of a wasm module as its public API.  i.e. we could inspect the wasm module at runtime an export the public functions out to the world of JS.   

It makes the actual API of the final wasm module self-describing.  Without this I can't tell what is the public API and what isn't, right?

tlively replied:
Ah, that seems reasonable. I'll add a configuration option for this.

tlively commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+  // Map internal names to (one of) their corresponding export names.
+  std::unordered_map<Name, Name> exports;
+  for (auto& ex : primary.exports) {
+    if (ex->kind != ExternalKind::Function) {
```
That's an interesting idea! I think you're right that if the simplifyGlobals optimization has already run, then the globals won't exist in the first place, though. In that case, I wouldn't want to complicate the code for a case that won't be important in practice.

kripken commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+  // Create the secondary module and copy trivial properties.
+  auto secondary = std::make_unique<Module>();
+  secondary->features = primary.features;
+  secondary->hasFeaturesSection = primary.hasFeaturesSection;
```
these could be in a helper in `module-utils` perhaps (`copyModuleFlags` - or a better name, that implies just global flags and not the contents)?

tlively replied:
Maybe, but I'm not sure it would be used anywhere else.

kripken commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+  // Map internal names to (one of) their corresponding export names.
+  std::unordered_map<Name, Name> exports;
+  for (auto& ex : primary.exports) {
+    if (ex->kind != ExternalKind::Function) {
```
why not functions? no obvious reason so the reader is forced to look lower down, unless I'm missing something obvious.

tlively replied:
Functions are exported only as necessary because we expect there to be many of them compared to memories, tables, globals, and events. I'll add a comment.

tlively commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+std::unique_ptr<Module> splitFunctions(Module& primary, const Config& config) {
+  std::set<Name> secondaryFuncs;
+  for (auto& func : primary.functions) {
+    if (config.primaryFuncs.find(func->name) == config.primaryFuncs.end()) {
```
Yeah, I can change to `count` here. I usually prefer `find` for performance reasons, but you're right that it shouldn't matter in this case. It's a travesty that C++ doesn't have a simple `contains` method on its containers.

sbc100 replied:
Ha, I just realized the irony of so-called "containers" not having a "contains" method :)

kripken commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+    if (exportIt != exports.end()) {
+      secondaryItem.base = exportIt->second;
+    } else {
+      exportName = Names::getValidExportName(primary, exportName);
```
you can avoid the if and always call this function - it will emit the simple name if there is no collision.

tlively replied:
The `if` is there to re-use the existing export name if it exists rather than using the `exportName` parameter. I'll rename the parameter to `genericExportName` to make that clearer.

kripken commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+    secondary.addGlobal(std::move(secondaryGlobal));
+  };
+  ModuleUtils::iterImportedGlobals(primary, shareGlobal);
+  ModuleUtils::iterDefinedGlobals(primary, shareGlobal);
```
is there a reason to have a helper function and these two calls, and not just
```cpp
for (auto& global : module->globals)
```
to iterate over them all? Seems like we do the same for imported and defined ones anyhow.

kripken commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+  ModuleUtils::iterDefinedEvents(primary, shareEvent);
+}
+
+template<class F> void forEachElement(Table& table, F f) {
```
Walkers walk  independent entities, including Expressions and Module elements like Exports, but not "immediates" of an entity, like table segment elements, which is what this does I think.

tlively replied:
No, I think this is rather specific to this use case. Critically, this function assumes that all table segments have constant offsets, which is not something we should be assuming in general. This visitor is much more fine-grained than the normal walkers and visitors; the generic ones visit the table, not individual table slots.

kripken commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+
+// Return the first free table slot and the segment to which to append new
+// items, if it exists.
+Index getFirstFreeTableIndex(Table& table, Table::Segment** outSegment) {
```
can't this use `forEachElement`?

tlively replied:
That would be overkill, since this doesn't need to access individual table slots, just the segments. The similarity is unfortunate, though.

kripken commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+// their names to their (eventual) table indices.
+void moveFunctions(Module& primary,
+                   Module& secondary,
+                   const std::set<Name>& secondaryFuncs) {
```
it may be slightly confusing that the set passed here is the ones to split out, but in the test it is the ones to keep. but maybe not worth changing.

tlively replied:
The public function matches the tests, so I think this is ok. Initially I was going to have the public function take a set of functions to split out, but I got feedback that any automatic analysis was more likely to identify the list of functions to keep.

kripken reviewed (commented):
Very nice work!

tlively commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+    secondary.table.initial = table->initial;
+    secondary.table.max = table->max;
+    makeImportExport(*table, secondary.table, "table", ExternalKind::Table);
+  };
```
I wouldn't call this deliberate, per se. I don't know where I picked up this coding style, but I like it because it limits the scope in which functions are available to the scope in which they are used. It also means you have just one place to look for variable declarations (previously in the function) rather than two (the function and the enclosing class).

I do think I could break up the function a bit, though. I was also considering using an RAII-style context manager to manage the new table slot allocations. I'll refactor this a bit and see if it seems any cleaner.

dschuff replied:
You may have picked it up here; Binaryen has always used it extensively; and I suspect @kripken picked it up from JS.

tlively commented on src/ir/module-splitting.h:
```diff
@@ -0,0 +1,44 @@
+  Name placeholderNamespace = "placeholder";
+};
+
+std::unique_ptr<Module> splitFunctions(Module& primary, const Config& config);
```
The return value is the secondary module; there is no placeholder module. "Placeholder functions" are imported JS functions that can be arbitrary hooks used to respond to a secondary function being called before it is ready. Possible behaviors of placeholder functions could be:

 1) Throwing some sort of "ModuleNotLoaded" error
 2) Blocking on the download and instantiation of the secondary module, then calling the requested function.
 3) Lazily starting the download of the secondary module, then doing (2)
 4) Using asyncify or stack switching to return to the main loop while the secondary module finishes downloading.

tlively replied:
I'll add a comment about the return value as suggested.

kripken replied:
Ah, I misunderstood the terms before, sorry about that....

tlively commented on test/example/module-splitting.cpp:
```diff
@@ -0,0 +1,208 @@
+
+  std::cout << "After:\n";
+  WasmPrinter::printModule(primary.get());
+  std::cout << "Secondary:\n";
```
See previous comment.

tlively commented on test/example/module-splitting.cpp:
```diff
@@ -0,0 +1,208 @@
+     )
+    ))");
+
+  // Non-deferred function
```
Sounds good.

tlively commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+  exportImportPrimaryFunctions(
+    primary, secondary, config.primaryFuncs, config.importNamespace);
+  setupTablePatching(
+    primary, secondary, secondaryFuncs, config.placeholderNamespace);
```
Possibly, I'll play around with it.

aheejin commented on test/example/module-splitting.txt:
```diff
@@ -0,0 +1,448 @@
+(module
+ (type $i32_=>_i32 (func (param i32) (result i32)))
+ (table $0 1 funcref)
+ (elem (i32.const 0) $foo)
```
This is not the same table with the one in the primary module, right? Why do we need the table in the secondary module here and why `$foo` has to be at address 0?

aheejin commented on test/example/module-splitting.txt:
```diff
@@ -0,0 +1,448 @@
+ (elem (i32.const 42) $placeholder_42)
+ (export "foo" (func $foo))
+ (export "table" (table $table))
+ (func $foo (param $0 i32) (result i32)
```
But this seems to share the table with primary module..?

aheejin reviewed (commented):
Nice! Haven't done reading yet, but some questions first:
- Why do we need trampolines? In place of calling trampolines, can't we do indirect call to place holders instead?
- Who connects placeholders to the right functions in the secondary module?

tlively commented on test/example/module-splitting.cpp:
```diff
@@ -0,0 +1,208 @@
+  WasmValidator validator;
+  bool valid;
+
+  auto primary = parse(&module.front());
```
Unfortunately this won't work because `parse` actually modifies the string it's parsing, but `c_str` returns `const char*`. I looked a fixing this, but it was more work than would be appropriate for this PR.

aheejin commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+    }
+  });
+
+  // Create active table segments in the secondary module to patch in the
```
Can table segments ever be passive?

tlively replied:
Yes, although that hasn't been implemented in Binaryen yet because no one is using it.

aheejin commented on test/example/module-splitting.txt:
```diff
@@ -0,0 +1,448 @@
+Secondary:
+(module
+ (type $i32_=>_i32 (func (param i32) (result i32)))
+ (table $0 1 funcref)
```
In case the primary module doesn't originally have a table but gets one during the module splitting process, the table is not exported, and it seems here the primary and the secondary module don't share the table and they end up with separate tables. Then it looks, how is the secondary module supposed to replace the table entries?

tlively replied:
Thanks for catching this. This is a bug that I missed in my first readthrough of the test output. The problem was that the table was created after the code to share it had already run. The fix is to move `shareImportableItems` to run last rather than first.

tlively commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+      };
+      CallCollector(primaryFuncs, calledPrimaryFuncs).walkFunction(func);
+    });
+  std::set<Name> calledPrimaryFuncs;
```
Hmm, I guess not, since it will be collected into another set later 👍

tlively replied:
I believe this comment originally applied to a different set, which is now a `std::vector` above.

aheejin replied:
Oh sorry, I was confused. It needs to be a set.

tlively commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+      return indexIt->second;
+    } else {
+      Index newIndex = firstFreeIndex + newTableElems.size();
+      newTableElems.push_back(secondaryFunc);
```
It seems cleaner to do the patching all at once at the end once all the new table slots have been allocated.

tlively commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+      *outSegment = &segment;
+    }
+  }
+  return firstFreeIndex;
```
Correct, this lets us just collect all the new elements in a single continuous vector that can be appended to an existing segment or turned into a new segment later. I'll add a comment.

tlively commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+
+  auto shareGlobal = [&](Global* global) {
+    assert(primary.features.hasMutableGlobals() &&
+           "TODO: add wrapper functions for disallowed mutable globals");
```
Since the globals are exported and imported, we need the mutable-globals feature to be enabled if there are any mutable globals in the module. I guess we technically don't need this assert for modules that don't have any mutable globals, but that seems niche.

aheejin replied:
Sorry I'm not sure if I understand. If a primary module doesn't have mutable globals, both primary and secondary modules will not need them, so everything is fine. If the primary module originally has mutable globals enabled, the secondary module will copy all those features, and both of them will have mutable globals, and everything's fine here. Why do we need to require anything about mutable globals for module splitting? Why is the mutable globals feature treated specially unlike all other features?

tlively replied:
The `mutable-globals` feature is a bit of a misnomer. What this feature actually allows is _imports and exports_ of mutable globals. It's totally valid to have mutable globals without it in MVP, but not to import or export them. So we need to check this feature here because importing or exporting a mutable global that was not previously imported or exported introduces a new use of the `mutable-globals` feature that was not there before.

aheejin replied:
I see... I was seriously confused about what mutable globals feature does. Thanks for pointing that out.

tlively replied:
Yeah the feature name really gives the wrong idea 😅

tlively commented on src/ir/module-splitting.h:
```diff
@@ -0,0 +1,44 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
I'll beef up the documentation of the functionality and assumptions in the header and copy most of the rest of the description including the description of the process into the .cpp file.

tlively commented:
Thank you all for the reviews so far! I've addressed all of the feedback (or at least I've meant to; let me know if I've missed something), and I've significantly refactored the code to break up the monolithic `moveFunctions` function. I think the code is much simpler and more readable now. I also fixed some gaps in the test suite and made the ordering of tests more principled.

tlively commented on src/ir/module-splitting.cpp:
```diff
@@ -0,0 +1,405 @@
+    const std::set<Name>& secondaryFuncs;
+    const std::map<Name, Signature>& secondarySignatures;
+    std::function<Index(Name)> getIndex;
+    Builder builder;
```
I ended up constructing the Builder rather than passing it in, so this no longer applies.

kripken reviewed (approved):
Looks good enough to land to me (for a completely new functionality I think it's more efficient to land it and iterate from there, so I didn't hunt for bugs on every single line - but if others want to review more carefully before landing I don't object of course).

tlively commented:
Thanks, @kripken! I'll land this tomorrow afternoon unless anyone has any new comments in the meantime.
